File: apps/backend/app/api/routes/analytics.py
"""
Analytics and reporting routes.

Dashboard summary, revenue reports, fleet utilization, fuel cost summary.
"""

from fastapi import APIRouter

from app.api.deps import CurrentUser, DbDep
from app.services import analytics_service

router = APIRouter(prefix="/analytics", tags=["analytics"])


@router.get("/dashboard")
async def get_dashboard(db: DbDep, user: CurrentUser):
    """Get high-level dashboard metrics."""
    return await analytics_service.dashboard_summary(db, user.organization_id)


@router.get("/revenue")
async def get_revenue_report(
    db: DbDep,
    user: CurrentUser,
    period: str = "monthly",
    months_back: int = 6,
):
    """Revenue breakdown by period (weekly or monthly)."""
    if period not in ("weekly", "monthly"):
        period = "monthly"
    if months_back < 1:
        months_back = 1
    elif months_back > 24:
        months_back = 24
    return await analytics_service.revenue_by_period(db, user.organization_id, period, months_back)


@router.get("/fleet-utilization")
async def get_fleet_utilization(db: DbDep, user: CurrentUser):
    """Fleet utilization report: per-driver load count, revenue, miles, current status."""
    return await analytics_service.fleet_utilization(db, user.organization_id)


@router.get("/fuel-costs")
async def get_fuel_costs(
    db: DbDep,
    user: CurrentUser,
    months_back: int = 3,
):
    """Fuel cost summary: total gallons, total cost, avg price per gallon."""
    if months_back < 1:
        months_back = 1
    elif months_back > 24:
        months_back = 24
    return await analytics_service.fuel_cost_summary(db, user.organization_id, months_back)

File: apps/openclaw/src/tools/analytics.ts
import { Type } from "@sinclair/typebox";
import type { AnyAgentTool } from "openclaw/plugin-sdk/core";
import { api, ok, err } from "../client.js";

export const tmsDashboard: AnyAgentTool = {
  name: "tms_dashboard",
  label: "TMS: Dashboard Summary",
  description:
    "Get a high-level KPI summary: active loads, total revenue, fleet size, and key metrics. " +
    "Use for quick status checks like 'how's business?' or 'what's our current load count?'",
  parameters: Type.Object({}),
  execute: async (_id) => {
    try {
      const data = await api.get("/api/analytics/dashboard");
      return ok(data);
    } catch (e) {
      return err(e);
    }
  },
};

export const tmsRevenueReport: AnyAgentTool = {
  name: "tms_revenue_report",
  label: "TMS: Revenue Report",
  description:
    "Get revenue broken down by time period. Use for questions like " +
    "'what did we make last month?' or 'show me weekly revenue trends'.",
  parameters: Type.Object({
    period: Type.Optional(
      Type.Union([Type.Literal("weekly"), Type.Literal("monthly")], {
        description: "Group by week or month (default: monthly)",
      }),
    ),
    months_back: Type.Optional(
      Type.Number({ description: "How many months of history to include (1-24, default 6)" }),
    ),
  }),
  execute: async (_id, params) => {
    try {
      const qs = new URLSearchParams();
      if (params.period) qs.set("period", params.period);
      if (params.months_back) qs.set("months_back", String(params.months_back));
      const data = await api.get(`/api/analytics/revenue?${qs}`);
      return ok(data);
    } catch (e) {
      return err(e);
    }
  },
};

export const tmsFleetUtilization: AnyAgentTool = {
  name: "tms_fleet_utilization",
  label: "TMS: Fleet Utilization",
  description:
    "Get per-driver utilization metrics: load count, total revenue, total miles, and current status. " +
    "Use for driver performance comparisons or answering 'who's been busiest?'",
  parameters: Type.Object({}),
  execute: async (_id) => {
    try {
      const data = await api.get("/api/analytics/fleet-utilization");
      return ok(data);
    } catch (e) {
      return err(e);
    }
  },
};

export const tmsFuelCosts: AnyAgentTool = {
  name: "tms_fuel_costs",
  label: "TMS: Fuel Cost Summary",
  description:
    "Get fuel spending summary over recent months. Use for expense tracking or answering " +
    "'how much did we spend on fuel last quarter?'",
  parameters: Type.Object({
    months_back: Type.Optional(
      Type.Number({ description: "How many months of history (1-24, default 3)" }),
    ),
  }),
  execute: async (_id, params) => {
    try {
      const qs = new URLSearchParams();
      if (params.months_back) qs.set("months_back", String(params.months_back));
      const data = await api.get(`/api/analytics/fuel-costs?${qs}`);
      return ok(data);
    } catch (e) {
      return err(e);
    }
  },
};
